test/core/end2end/cq_verifier.h

```diff
@@ -64,10 +65,19 @@ class CqVerifier {
   using ExpectedResult =
       absl::variant<bool, Maybe, AnyStatus, PerformAction, MaybePerformAction>;
 
+  // Captures information about one failure
   struct Failure {
     SourceLocation location;
     std::string message;
     std::vector<std::string> expected;
+    std::vector<std::string> message_details;
+  };
+
+  // Produces a string upon the successful (but unexpected) completion of an
+  // expectation.
+  class SuccessfulStateString {
+   public:
+    virtual std::string GetSuccessfulStateString() = 0;
```

Comment:
needs a virtual destructor.

Reply:
No it doesn't, and we should not give this blanket advice.

It needs a virtual destructor *or* a non-virtual protected destructor, and for gRPC we should always prefer the latter - no need to add an indirection to destruction unnecessarily.

Added a protected destructor.